src/core/algorithms/dd/split/split.cpp

```diff
@@ -0,0 +1,702 @@
+                        if (!is_removable) remove = true;
+                        is_removable = true;
+                        break;
+                    }
+                }
+                if (is_removable) break;
+            }
+            if (!remove) results_copy.push_back(dd3);
+        }
+        if (results_copy.size() == results_.size()) break;
+        results_ = results_copy;
+    }
+    return num_cycles;
+}
+
+double Split::CalculateDistance(std::size_t column_index,
+                                std::pair<std::size_t, std::size_t> tuple_pair) {
+    model::TypedColumnData const& column = typed_relation_->GetColumnData(column_index);
+    model::TypeId type_id = column.GetTypeId();
+
+    if (type_id == +model::TypeId::kUndefined) {
+        throw std::invalid_argument("Column with index \"" + std::to_string(column_index) +
+                                    "\" type undefined.");
+    }
+    if (type_id == +model::TypeId::kMixed) {
+        throw std::invalid_argument("Column with index \"" + std::to_string(column_index) +
+                                    "\" contains values of different types.");
+    }
+    if (column.IsNull(tuple_pair.first) || column.IsNull(tuple_pair.second)) {
+        throw std::runtime_error("Some of the value coordinates are nulls.");
+    }
+    if (column.IsEmpty(tuple_pair.first) || column.IsEmpty(tuple_pair.second)) {
+        throw std::runtime_error("Some of the value coordinates are empties.");
+    }
+    double dif = 0;
+    if (column.GetType().IsMetrizable()) {
+        std::byte const* first_value = column.GetValue(tuple_pair.first);
+        std::byte const* second_value = column.GetValue(tuple_pair.second);
+        auto const& type = static_cast<model::IMetrizableType const&>(column.GetType());
+        dif = type.Dist(first_value, second_value);
+    }
+    return dif;
+}
+
+// must be inline for optimization (gcc 11.4.0)
+inline bool Split::CheckDF(df const& dif_func, std::pair<std::size_t, std::size_t> tuple_pair) {
+    for (std::size_t column_index = 0; column_index < num_columns_; column_index++) {
+        double dif = distances_[column_index][tuple_pair.first][tuple_pair.second];
+        if (dif < dif_func[column_index].first || dif > dif_func[column_index].second) {
+            return false;
+        }
+    }
+    return true;
+}
+
+bool Split::VerifyDD(dd const& dep) {
+    for (std::size_t i = 0; i < num_rows_; i++) {
+        for (std::size_t j = i + 1; j < num_rows_; j++) {
+            if (CheckDF(dep.lhs, {i, j}) && !CheckDF(dep.rhs, {i, j})) return false;
+        }
+    }
+    return true;
+}
+
+void Split::CalculateAllDistances() {
+    for (std::size_t column_index = 0; column_index < num_columns_; column_index++) {
+        distances_.push_back({});
+
+        std::shared_ptr<model::PLI const> pli =
+                relation_->GetColumnData(column_index).GetPliOwnership();
+        std::deque<model::PLI::Cluster> const& index = pli->GetIndex();
```

Comment:
I use `position_list_index` to obtain clusters of equal values for each column. `index` contains lists of row numbers for each cluster and `pt` stores numbers of clusters for each row. This helps to reduce the number of distance calculations because if values are in the same cluster then the distance between them is `0`, and for any two values from two fixed clusters the distance is the same. In the first loop I calculate distances between clusters and in the second loop for the remaining values.